0004-remove-spandata.md

```diff
@@ -0,0 +1,51 @@
+# Remove SpanData
+
+Remove and replace SpanData by adding span start and finish options.
+
+## Motivation
+
+SpanData represents an immutable span object, creating a fairly large API for all of the fields (12 to be exact). It exposes what feels like an SDK concern and implementation detail to the API surface. As a user, this is another API I need to learn how to use, and ID generation might also need to be exposed. As an implementer, it is a new data type that needs to be supported. The primary motivation for removing SpanData revolves around the desire to reduce the size of the tracing API.
+
+## Explanation
+
+SpanData has a couple of use cases.
+
+The first use case revolves around creating a span syncronously but needing to change the start time to a more accurate timestamp. For example, in an HTTP server, you might record the time the first byte was received, parse the headers, determine the span name, and then create the span. The moment the span was created isn't representative of when the request actually began, so the time the first byte was received would become the span's start time. Since the current API doesn't allow start timestamps, you'd need to create a SpanData object. The big downside is that you don't end up with an active span object.
+
+The second use case comes from the need to construct and report out of band spans, meaning that you're creating "custom" spans for an operation you don't own. One good example of this is a span sink that takes in structured logs that contain correlation IDs and a duration (e.g. from splunk) and [converts them](https://github.com/lightstep/splunktospan/blob/master/splunktospan/span.py#L43) to spans for your tracing system. [Another example](https://github.com/lightstep/haproxy_log2span/blob/master/lib/lib.go#L292) is running a sidecar on an HAProxy machine, tailing the request logs, and creating spans. SpanData allows you to report the out of band reporting case, whereas you can’t with the current Span API as you cannot set the start and end timestamp.
+
+I'd like to propose getting rid of SpanData and `tracer.recordSpanData()` and replacing it by allowing `tracer.startSpan()` to accept a start timestamp option and `span.finish()` to accept end timestamp option. This reduces the API surface, consolidating on a single span type. Options would meet the requirements for out of band reporting.
+
+## Internal details
+
+`startSpan()` would change so you can include options such as `withStartTimestamp()` and `withResource()`. When you have a span sink, out of band spans may have different resources than the tracer they are being reported to, so you want to pass an explicit resource. For `span.finish()` you would have `withEndTimestamp()`. The exact implementation would be language specific, so some would use an options pattern with function overloading or variadic parameters, or add these options to the span builder.
```

Comment:
I don't like this aspect of the OpenTracing API since the clock resolution of an external timestamp is not as good as the cpu "nano" time.  I suggest we expose a method to get the "nano" time from the same clock used by the tracer (which perhaps returns a "typed" value instead of just a primitive), and accept the return value of that as the start/stop time argument in the builder.

The description in this document is pretty generic and likely to result in significant variation in implementation.  Please clarify

Reply:
I think this comment applies when we write the specs for this RFC. We should keep this RFC a bit more high level.

Happy to hear what others think.

Reply:
Fair point.  Consider this feedback for the future spec. Carry on.